src/utils/intrusive_ptr_test.cc

```diff
@@ -0,0 +1,155 @@
+//
+//     https://www.apache.org/licenses/LICENSE-2.0
+//
+// Unless required by applicable law or agreed to in writing, software
+// distributed under the License is distributed on an "AS IS" BASIS,
+// WITHOUT WARRANTIES OR CONDITIONS OF ANY KIND, either express or implied.
+// See the License for the specific language governing permissions and
+// limitations under the License.
+//-----------------------------------------------------------------------------
+#include "src/utils/intrusive_ptr.h"
+
+#include <iostream>
+#include <string>
+
+#include "src/common/testing/gtest.h"
+
+namespace raksha {
+namespace {
+
+class RefCountedType {
+ public:
+  RefCountedType(std::string id = "<unknown>", unsigned int initial_count = 0)
+      : id_(id), count_(initial_count) {}
+
+  unsigned int count() const { return count_; }
+  void set_id(std::string id) { std::swap(id_, id); }
+  const std::string& id() const { return id_; }
+  void Retain() { count_++; }
+  void Release() { count_--; }
+
+ private:
+  std::string id_;
+  unsigned int count_;
+};
+
+using RefCountedTypePtr = intrusive_ptr<RefCountedType>;
+
+TEST(IntrusivePtrTest, DefaultConstructorInitializesToNullptr) {
+  RefCountedTypePtr null;
+  EXPECT_EQ(null.get(), nullptr);
+}
+
+TEST(IntrusivePtrTest, ConstructorIncreasesCountAndSavesPtr) {
+  auto value = std::make_unique<RefCountedType>();
+  EXPECT_EQ(value->count(), 0);
+
+  RefCountedTypePtr ptr(value.get());
+  EXPECT_EQ(value->count(), 1);
+
+  EXPECT_EQ(ptr.get(), value.get());
+}
+
+TEST(IntrusivePtrTest, CopyConstructorIncreasesCountAndSharesPtr) {
+  auto value = std::make_unique<RefCountedType>();
+  EXPECT_EQ(value->count(), 0);
+
+  RefCountedTypePtr ptr(value.get());
+  EXPECT_EQ(value->count(), 1);
+
+  RefCountedTypePtr copy = ptr;
+  EXPECT_EQ(value->count(), 2);
+  EXPECT_EQ(ptr.get(), copy.get());
+}
+
+TEST(IntrusivePtrTest, MoveConstructorMaintainsCountAndNullifiesOther) {
+  auto value = std::make_unique<RefCountedType>();
+  EXPECT_EQ(value->count(), 0);
+
+  RefCountedTypePtr ptr(value.get());
+  EXPECT_EQ(value->count(), 1);
+
+  RefCountedTypePtr moved = std::move(ptr);
+  EXPECT_EQ(value->count(), 1);
+  EXPECT_EQ(ptr.get(), nullptr);
+}
+
+TEST(IntrusivePtrTest, AssignmentIncreasesCountAndSharesPtr) {
+  auto value = std::make_unique<RefCountedType>();
+  EXPECT_EQ(value->count(), 0);
+
+  RefCountedTypePtr ptr(value.get());
+  EXPECT_EQ(value->count(), 1);
+
+  RefCountedTypePtr assigned;
+  EXPECT_EQ(assigned.get(), nullptr);
+
+  assigned = ptr;
+  EXPECT_EQ(value->count(), 2);
+  EXPECT_EQ(ptr.get(), assigned.get());
+}
+
+TEST(IntrusivePtrTest, GetReturnsSavedPointer) {
+  auto value = std::make_unique<RefCountedType>();
+  RefCountedTypePtr ptr(value.get());
+  EXPECT_EQ(ptr.get(), value.get());
+}
+
+TEST(IntrusivePtrTest, DereferenceReturnsSavedObject) {
+  auto value = std::make_unique<RefCountedType>();
+  RefCountedTypePtr ptr(value.get());
+  EXPECT_EQ(&(*ptr), value.get());
+}
```

Comment:
Similarly, is there a double free here?

Reply:
Refactored the code, but this would not have caused a double free earlier. In the previous version,  the custom implementation of `Retain` and `Release` only kept track of ref counts and did not the object. It is changed now.